clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -181,33 +256,465 @@ public void updateState(ConsumerGroupHeartbeatResponseData response) {
+            if (metadataError != null) {
+                log.error("Reconciliation failed due to error getting metadata to resolve topic " +
+                        "names for topic IDs {} in target assignment.", targetAssignment);
+                // TODO: failing reconciliation (no ack sent to the broker), but leaving
+                //  member in STABLE state. Double check if any other action should be taken
+                //  here.
+                transitionTo(MemberState.STABLE);
+                return;
+            }
+
+            if (!subscriptions.checkAssignmentMatchedSubscription(assignedPartitions)) {
+                log.debug("Target assignment {} does not match the current subscription {}; it is " +
+                                "likely that the subscription has changed since we joined the group, " +
+                                "will re-join with current subscription",
+                        targetAssignment,
+                        subscriptions.prettyString());
+                transitionToJoining();
+                return;
+            }
+
+            // Partitions to assign (not previously owned)
+            SortedSet<TopicPartition> addedPartitions = new TreeSet<>(COMPARATOR);
+            addedPartitions.addAll(assignedPartitions);
+            addedPartitions.removeAll(ownedPartitions);
+
+
+            // Partitions to revoke
+            SortedSet<TopicPartition> revokedPartitions = new TreeSet<>(COMPARATOR);
+            revokedPartitions.addAll(ownedPartitions);
+            revokedPartitions.removeAll(assignedPartitions);
+
+            log.info("Updating assignment with\n" +
+                            "\tAssigned partitions:                       {}\n" +
+                            "\tCurrent owned partitions:                  {}\n" +
+                            "\tAdded partitions (assigned - owned):       {}\n" +
+                            "\tRevoked partitions (owned - assigned):     {}\n",
+                    assignedPartitions,
+                    ownedPartitions,
+                    addedPartitions,
+                    revokedPartitions
+            );
+
+            CompletableFuture<Void> revocationResult;
+            if (!revokedPartitions.isEmpty()) {
+                revocationResult = revokePartitions(revokedPartitions);
+            } else {
+                revocationResult = CompletableFuture.completedFuture(null);
+                // Reschedule the auto commit starting from now (new assignment received without any
+                // revocation).
+                commitRequestManager.resetAutoCommitTimer();
+            }
+
+            // Future that will complete when the full reconciliation process completes (revocation
+            // and assignment, executed sequentially)
+            CompletableFuture<Void> reconciliationResult =
+                    revocationResult.thenCompose(r -> {
+                        if (state == MemberState.RECONCILING) {
```

Comment:
Totally valid point. I fixed it, but comparing assignments seemed more complicated and harder to reason about with the current approach. Now the target assignment is kind of a moving target, it can be modified anytime not only from the server, but also from metadata updates. 

So back to the problem of making sure that delayed reconciliations are not applied after a member rejoins, I added just a check based on the current member ID, to identify that a reconciliation completed but the member is already re-joining. What do you think? (whenever there are no rejoins in the picture, I expect that the RECONCILING state check alone should be enough given that reconciliations are always applied sequentially)